File: 125.valid-palindrome.py
# @leet start
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = s.replace("/[^0-9A-Za-z]/g", "").lower()
        l=0
        r= len(s)
        while l<r:
            if s[l] != s[r]:
                return False
            l++
            r--
        return True




# @leet end

File: 125.valid-palindrome.ts
// @leet start
function isPalindrome(s: string): boolean {
  let l = 0;
  let r = s.length - 1;
  const isAlphaNum = (c: number) =>
    (c >= 48 && c <= 57) ||    // 0-9
    (c >= 65 && c <= 90) ||    // A-Z
    (c >= 97 && c <= 122);     // a-z
  const toLowerCode = (c: number) =>
    (c >= 65 && c <= 90) ? (c + 32) : c; // A-Z -> a-z
  while (l < r) {
    let cl = s.charCodeAt(l);
    let cr = s.charCodeAt(r);
    while (l < r && !isAlphaNum(cl)) cl = s.charCodeAt(++l);
    while (l < r && !isAlphaNum(cr)) cr = s.charCodeAt(--r);
    if (toLowerCode(cl) !== toLowerCode(cr)) return false;
    l++;
    r--;
  }
  return true;
}
// @leet end
